contrib/gitlab/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,144 @@
+)
+
+const (
+	GLHeaderEvent    = "GitLab-Event"
+	GLHeaderDelivery = "GitLab-Delivery"
+)
+
+// Adapter converts incoming GitLab webhook events to CloudEvents
+type Adapter struct {
+	client client.Client
+}
+
+// New creates an adapter to convert incoming GitLab webhook events to CloudEvents and
+// then sends them to the specified Sink
+func New(sinkURI string) (*Adapter, error) {
+	a := new(Adapter)
+	var err error
+	a.client, err = kncloudevents.NewDefaultClient(sinkURI)
+	if err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+// HandleEvent is invoked whenever an event comes in from GitLab
+func (a *Adapter) HandleEvent(payload interface{}, header http.Header) {
+	hdr := http.Header(header)
+	err := a.handleEvent(payload, hdr)
+	if err != nil {
+		log.Printf("unexpected error handling GitLab event: %s", err)
+	}
+}
+
+func (a *Adapter) handleEvent(payload interface{}, hdr http.Header) error {
+	gitLabEventType := hdr.Get("X-" + GLHeaderEvent)
+	eventID := hdr.Get("X-" + GLHeaderDelivery)
+	extensions := map[string]interface{}{
+		GLHeaderEvent:    gitLabEventType,
+		GLHeaderDelivery: eventID,
+	}
+
+	log.Printf("Handling %s", gitLabEventType)
+
+	cloudEventType := fmt.Sprintf("%s.%s", sourcesv1alpha1.GitLabSourceEventPrefix, gitLabEventType)
+	source, err := sourceFromGitLabEvent(gl.Event(gitLabEventType), payload)
+	if err != nil {
+		return err
+	}
+
+	event := cloudevents.Event{
+		Context: cloudevents.EventContextV02{
+			ID:         eventID,
+			Type:       cloudEventType,
+			Source:     *source,
+			Extensions: extensions,
+		}.AsV02(),
+		Data: payload,
+	}
+	_, err = a.client.Send(context.TODO(), event)
+	return err
+}
+
+func sourceFromGitLabEvent(gitLabEvent gl.Event, payload interface{}) (*types.URLRef, error) {
+	var url string
+	switch gitLabEvent {
+	case gl.PushEvents:
+		pe := payload.(gl.PushEventPayload)
+		url = pe.Repository.URL
+	case gl.TagEvents:
+		te := payload.(gl.TagEventPayload)
+		url = te.Repository.URL
+	case gl.IssuesEvents:
+		ie := payload.(gl.IssueEventPayload)
+		url = ie.ObjectAttributes.URL
+	case gl.ConfidentialIssuesEvents:
+		cie := payload.(gl.ConfidentialIssueEventPayload)
+		url = cie.ObjectAttributes.URL
+	case gl.CommentEvents:
+		ce := payload.(gl.CommentEventPayload)
+		url = ce.ObjectAttributes.URL
+	case gl.MergeRequestEvents:
+		mre := payload.(gl.MergeRequestEventPayload)
+		url = mre.ObjectAttributes.URL
+	case gl.WikiPageEvents:
+		wpe := payload.(gl.WikiPageEventPayload)
+		url = wpe.ObjectAttributes.URL
+	case gl.PipelineEvents:
+		pe := payload.(gl.PipelineEventPayload)
+		url = pe.ObjectAttributes.URL
+	case gl.BuildEvents:
+		be := payload.(gl.BuildEventPayload)
+		url = be.Repository.URL
+	}
+
+	log.Printf("url %s", url)
+	//Hack here, since gitlab/githab could use ssh url. it's not typical URL style, like this: git@github.com:knative/eventing-sources.git
+	//I guess it could be setting in the gitlab/github, but leave the code here to prevent exception
+	if url != "" {
```

Comment:
Can you explain in which cases this problem can occur? We do have `git_http_url` in a few objects, but I see that we are not using it here :thinking:

Reply:
If use ssh url like "git@github.com:tektoncd/pipeline.git", it's not a typical url which could parse by `net/url` , so I hack it 😄